### ISSUE TITLE
Question about logging

### ISSUE THREAD

**Issue by Michael-F-Ellis** [bottlepy/bottle#985]
**TL;DR**: I've a python logging configuration I've used for years in multiprocessing applications. How can I make it work in a bottle app spawned as one of the processes in my application?

**More info:**
With multiple processes running, it's really useful to have all the log messages from every process go to a common file in a format that includes a timestamp, the module name, etc.  It's also useful to map logger.debug(), logger.info(), etc to module level names like debug(), info(), etc.  

**Example**
In modules containing spawned processes, I init the logger like so:
```
import local_logging
debug, info, warning, error = local_logging.init_logger(msg_prefix="Whatever") 
```
and use the returned functions to create log messages.
```
info('Entering serial_response_handler_service')
```
produces a message like this
```
2017-07-14 09:54:35,064 - pid:39690 in:serial_response_handler.serial_response_handler_service() - INFO
Whatever: Entering serial_response_handler_service
```
**The Problem**
When I try this in a bottle application, it only works within the function that launches the app.  Trying to use it within a callback doesn't create an error, but it doesn't write the message to the log file. I've no idea where it's going.

_The logging call works here_
```
def serve(const, server='wsgiref', port=8001, reloader=False,
          debugmode=False, quiet=True):

    info("starting web server")
    bottle.debug(debugmode) ##

    ## Launch the web service loop.
    bottle.run(app,
               host='0.0.0.0',
               server=server,
               port=port,
               reloader=reloader,
               debug=debugmode,
               quiet=quiet)         
```
_but not here_

```
@app.post("/checktoken")
def checktoken(db):
    info("Checking token")
    token = request.forms.get("token")
    return dict(token_valid=auth.checktoken(token, db))   
```

I know there are various logging plugins and solutions on SO but I'd really like to keep my familiar way of writing log messages and have it work in callbacks. 

I'm appending the code in my logging module at the end of this message. As you can see, it's just a configuration of the standard Python logging facility.  Any help in figuring out to make it work with Bottle will be much appreciated.

Thanks,
Mike Ellis

```
"""
Simple general purpose logger.
Usage: in __main__ only, place the following line at module scope
before calling init_logger:

sys.modules['__main__']['LOG_FILENAME'] = "path/to/your/logfile"

In all modules (__main__ included) initialize the logger as follows:
debug, info, warning, error = logtest.init_logger()

Then you can freely call any of the above four procedures to create log
messages at the desired level.  The logfiles are rotated daily at midnight
with 7 days retention.

"""

import sys
import os
import logging
import logging.handlers
import atexit
from traceback import format_exception
from multiprocessing import current_process
def on_exit():
    """ Flush the logs when the program exits. """
    logging.shutdown()

atexit.register(on_exit)

DEFAULT = 0
USE_EXISTING = 1
CLEAR_EXISTING = 2

_logger = None  # global to hold the configured loggger.
def init_logger(action = DEFAULT, msg_prefix="", logfile=None):
    """ Call this once in your module to initialize our logger settings. """
    if logfile:
        filename=logfile
    else:
        try:
            filename = sys.modules['__main__'].LOG_FILENAME
        except AttributeError:
            ## typically needed for unit testing
            filename = "/tmp/default.log"

    # Set up a specific logger with our desired output level
    #pid = os.getpid()
    #my_logger = logging.getLogger(str(pid))
    my_logger = logging.getLogger()  # Use the root logger
    my_logger.setLevel(logging.DEBUG)

    #if action == CLEAR_EXISTING:
    #    # Use in forked processes prevent duplicate messages.
    #    nhandlers = len(my_logger.handlers)
    #    for i in xrange(nhandlers):
    #        my_logger.handlers.pop(0)

    #elif action != USE_EXISTING:
    if action != USE_EXISTING:
        # Use in the __main__ module to create a handler.
        handler = logging.handlers.TimedRotatingFileHandler(
            filename=filename,
            when='midnight', interval=1, backupCount=7)

        formatter = logging.Formatter('%(asctime)s - pid:%(process)s - '
                                      'in:%(module)s.%(funcName)s() - '
                                      '%(levelname)s'
                                      '\n%(processName)s: %(message)s')
        handler.setFormatter(formatter)
        my_logger.addHandler(handler)

    ## Try always clearing all but last in list
    nhandlers = len(my_logger.handlers)
    for _ in range(nhandlers - 1):
        my_logger.handlers.pop(0)


    def logger_excepthook(etype, value, traceback):
        """ See python docs for sys.excepthook to understand how this works. """
        my_logger.error("".join(format_exception(etype, value, traceback)))

    sys.excepthook = logger_excepthook

    global _logger
    _logger = my_logger
    #return debug, info, warning, error
    return my_logger.debug, my_logger.info, my_logger.warning, my_logger.error

def setLevel(level):
    """ So we can change it locally """
    ## Accept lowercase string names for levels
    levels = dict(
        debug=logging.DEBUG,
        info=logging.INFO,
        warning=logging.WARNING,
        error=logging.ERROR
        )
    try:
        _logger.setLevel(levels[level])
        return True
    except KeyError:
        return False

if __name__ == "__main__":
    # """ For unit testing """
    LOG_FILENAME = "/tmp/junk.log"
    debug, info, warning, error = init_logger()
    debug("debug message")
    info("info message")
    warning("warning message")
    ## deliberate uncaught error
    _x = 1/0

```

**Comment by defnull** [comment]
I have no idea. Which server back-end are you using? If your server back-end spawns multiple processes, and they all access the same file, there might be race conditions on the logfile. Try logging to stdout instead of a file and see if that helps?

**Comment by Michael-F-Ellis** [comment]
Same result with the default wsgi server and cherrypy.  As to race conditions, I've been using this configuration for years with multiple processes and have never encountered anything that seemed to be a race condition. I should clarify, in case it matters, that the server is not the parent process in my app.  It's launched along with the other processes from a common parent.

**Comment by defnull** [comment]
Bottle does not do anything special with handler callbacks. If `info` is what you think it is (and not somehow overwritten or shadowed by a different variable in global or closure scope), then it is just called as a normal function from whatever thread bottle currently runs in. No magic.

**Comment by Michael-F-Ellis** [comment]
> No magic.

That's exactly why I like Bottle.  (and why I gave up on Web2Py).

I'll try to create a minimal example to see if I can reproduce the problem.

**Comment by Michael-F-Ellis** [comment]
It suddenly started working.  Programmer brain damage, I suppose.   Apologies for the noise ..

**Comment by defnull** [comment]
No prob :) It happens.
